components/raftstore/src/store/peer_storage.rs

```diff
@@ -757,7 +796,116 @@ where
         Ok(())
     }
 
-    pub fn entries(&self, low: u64, high: u64, max_size: u64) -> raft::Result<Vec<Entry>> {
+    pub fn set_async_fetch_res(&mut self, res: RaftlogFetchResult) {
+        let prev_res = self.async_fetch_res.replace(Some(res));
+        debug_assert!(prev_res.is_none());
+        if let Some(r) = prev_res {
+            warn!(
+                "unexpected async fetch res";
+                "region_id" => self.region.get_id(),
+                "peer_id" => self.peer_id,
+                "res" => ?r,
+            );
+        }
+    }
+
+    fn async_fetch(
+        &self,
+        region_id: u64,
+        low: u64,
+        high: u64,
+        max_size: u64,
+        context: GetEntriesContext,
+        buf: &mut Vec<Entry>,
+    ) -> raft::Result<usize> {
+        if let Some(res) = self.async_fetch_res.replace(None) {
+            if res.low == low {
+                let mut ents = res.ents?;
+                return Ok(match ents.first() {
+                    None => 0,
+                    Some(Entry { index: first, .. }) => {
+                        // no need to compare the index of last entry with high, it's okay to not return the full entries
```

Comment:
The result can be used only when `first index == low && result max size == max_size`.

Reply:
when the size of result < max_size, seems it doesn't make any problem.

Reply:
For sending logs, it may be OK. But it will not be OK in other cases. When caller calls `entries` with size big enough, then it expects to get all the entries between `low` and `high`. Returning partial results breaks the assumption of the API.

Reply:
if that it should also consider `high`. But it may cause async fetch results always invalid.

Reply:
seems we can give the assumption only when `max_size == None`?

Reply:
> if that it should also consider `high`. But it may cause async fetch results always be invalid.

Why? The key point is whether the old result is stopped at high limit or size limit. If it hits size limit, then as long as `low` and `size_limit` are the same, the result can be reused directly. If it stops at high limit, then it should either fetches more from cache or trigger a second fetch.